src/constbuffer.c

```diff
@@ -394,3 +398,256 @@ IMPLEMENT_MOCKABLE_FUNCTION(, bool, CONSTBUFFER_HANDLE_contain_same, CONSTBUFFER
+{
+    uint32_t result;
+    /*Codes_SRS_CONSTBUFFER_02_041: [ If source is NULL then CONSTBUFFER_get_serialization_size shall fail and return 0. ]*/
+    if (source == NULL)
+    {
+        LogError("invalid argument CONSTBUFFER_HANDLE source=%p", source);
+        result = 0;
+    }
+    else
+    {
+        /*Codes_SRS_CONSTBUFFER_02_042: [ If sizeof(uint8_t) + sizeof(uint32_t) + source's size exceed UINT32_MAX then CONSTBUFFER_get_serialization_size shall fail and return 0. ]*/
+        if (source->alias.size > UINT32_MAX - sizeof(uint8_t) - sizeof(uint32_t))
+        {
+            LogError("serialization size exceeds UINT32_MAX=%" PRIu32 ". It is the sum of sizeof(uint8_t)=%zu + sizeof(uint32_t)=%zu + source->alias.size=%zu",
+                UINT32_MAX, sizeof(uint8_t), sizeof(uint32_t), source->alias.size);
+            result = 0;
+        }
+        else
+        {
+            /*Codes_SRS_CONSTBUFFER_02_043: [ Otherwise CONSTBUFFER_get_serialization_size shall succeed and return sizeof(uint8_t) + sizeof(uint32_t) + source's size. ]*/
+            result = sizeof(uint8_t) + sizeof(uint32_t) + (uint32_t)source->alias.size;
+        }
+    }
+    return result;
+}
+
+/*THE_LESSER is the smallest size of the SIZE_MAX, UINT32_MAX. This is needed because 
+1) there's a mix of size_t (malloc) and uint32_t (usage) about sizes
+2) sizeof(size_t) is not a constant
+3) that is detrimental to any portable serialization.
+
+So the lesser of the two is chosen in the overflow computations*/
+#if SIZE_MAX > UINT32_MAX
+#define THE_LESSER UINT32_MAX
+#define PRI_THE_LESSER PRIu32
+typedef uint32_t THE_LESSER_T;
+#else
+#define THE_LESSER SIZE_MAX
+#define PRI_THE_LESSER "zu"
+typedef size_t THE_LESSER_T;
+#endif
+
+static void* calls_malloc(size_t size, void* context)
+{
+    (void)context;
+    return malloc(size);
+}
+
+unsigned char* CONSTBUFFER_to_buffer(CONSTBUFFER_HANDLE source, CONSTBUFFER_to_buffer_alloc alloc, void* alloc_context, uint32_t* size)
+{
+    unsigned char* result;
+    if (
+        /*Codes_SRS_CONSTBUFFER_02_044: [ If source is NULL then CONSTBUFFER_to_buffer shall fail and return NULL. ]*/
+        (source == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_045: [ If size is NULL then CONSTBUFFER_to_buffer shall fail and return NULL. ]*/
+        (size == NULL)
+        )
+    {
+        LogError("invalid arguments CONSTBUFFER_HANDLE source=%p, CONSTBUFFER_to_buffer_alloc alloc=%p, uint32_t* size=%p",
+            source, alloc, size);
+        result = NULL;
+    }
+    else
+    {
+        /*Codes_SRS_CONSTBUFFER_02_046: [ If alloc is NULL then CONSTBUFFER_to_buffer shall use malloc as provided by gballoc_hl_redirect.h. ]*/
+        if (alloc == NULL)
+        {
+            alloc = &calls_malloc;
+        }
+
+        /*Codes_SRS_CONSTBUFFER_02_054: [ If there are any failures then CONSTBUFFER_to_buffer shall fail and return NULL. ]*/
+        if (THE_LESSER - sizeof(uint8_t) - sizeof(uint32_t) < source->alias.size)
```

Comment:
>THE_LESSER [](http://example.com/codeflow?start=12&length=10)

Please file a task in the bucket of stuff that never gets done so that we start the fire for forging the statue of the unknown hero that will change size_t to uint32_t.

Reply:
Task 10783889: Change CONSTBUFFER_HANDLE internal size from size_t to uint32_t

[https://msazure.visualstudio.com/One/_workitems/edit/10783889](https://msazure.visualstudio.com/One/_workitems/edit/10783889)